cmd/skopeo/proxy.go

```diff
@@ -232,6 +233,56 @@ func (h *proxyHandler) OpenImage(args []interface{}) (replyBuf, error) {
 	return ret, nil
 }
 
+func isNotFoundImageError(err error) bool {
+	msg := err.Error()
+	// TODO: Drive a correct API for this into containers/image
+	return strings.HasSuffix(msg, "manifest unknown") ||
+		strings.Contains(msg, ": no descriptor found for reference")
+}
+
+// OpenImage accepts a string image reference i.e. TRANSPORT:REF - like `skopeo copy`.
+// The return value is an opaque integer handle.  If the image does not exist, zero
+// is returned.
+func (h *proxyHandler) OpenImageOptional(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	var ret replyBuf
+
+	if h.sysctx == nil {
+		return ret, fmt.Errorf("client error: must invoke Initialize")
+	}
+	if len(args) != 1 {
+		return ret, fmt.Errorf("invalid request, expecting one argument")
+	}
+	imageref, ok := args[0].(string)
+	if !ok {
+		return ret, fmt.Errorf("expecting string imageref, not %T", args[0])
+	}
+
+	imgRef, err := alltransports.ParseImageName(imageref)
+	if err != nil {
+		return ret, err
+	}
+	imgsrc, err := imgRef.NewImageSource(context.Background(), h.sysctx)
+	if err != nil {
+		if isNotFoundImageError(err) {
+			ret.value = 0
+			return ret, nil
+		}
+		return ret, err
+	}
+
+	h.imageSerial++
+	openimg := &openImage{
+		id:  h.imageSerial,
+		src: imgsrc,
+	}
+	h.images[openimg.id] = openimg
```

Comment:
This does not avoid the `ret.value = 0` value. Maybe factor ~this part out of the two `OpenImage` methods, and have it always skip `0`.

Reply:
I'm not sure I understand; note that we increment `h.imageSerial` before the success return path.  Oh, are you worried about overflow?  Getting up to 4 billion images seems...unlikely?

Reply:
Ugh ok actually this overflow thing gave me flashbacks of https://gitlab.gnome.org/GNOME/glib/-/commit/b98a1c8df30f9e24588a48331dacf01e49760549 so...hmm, I think we could bump it to a `uint64` actually...

Reply:
It’s a fair point that this is unlikely — actually if we did overflow, there is nothing in the existing code that avoids re-allocating a still used ID, so that would be broken anyway.

Really I was focusing on the _documented semantics_ of the method. If `0` means “image does not exist”, then the image ID space should clearly exclude `0`. Alternatively, instead of a reserved image ID value, add an extra `failedBecauseImageDidNotExist` field, and then the whole image ID space can be used without restrictions.

(It seemed to me that consolidating the “create `openImage` + return an ID” code was also a bit easier to maintain than the two copies, and adding three lines to explicitly skip `0` (or all non-positive values?) would have been cheap enough. But if we don’t avoid duplicates — and I agree that’s really unnecessary work at this point — then there’s not that much benefit.)

I guess I’d be fine with just a comment to the effect of “Note that if imageID ever overflows, we start allocating `0`, which has special meaning, and already-allocated values, and this completely falls apart.

---

int64 is a bit risky in JSON, where numbers are IIRC float with 53-bit precision unless users explicitly do some non-default parsing. Also note that the existing tests explicitly coerce to `uint32`.